### PR TITLE
Update robo to newer Orion version.

### PR DESCRIPTION
**Pull request by lebrice**
- Updates all algos to use the most recent version of Orion.
- Changes to the installation instructions (only one pip install step is required). 
- Temporarily disables the RandomForest model (dependency change needs to be fixed)
- Make tests a lot faster to run (DNGO in particular)
- Fix memory leak and innefficiencies in DNGO training loop.
- Add CUDA compatibility to DNGO

NOTE: This PR depends on https://github.com/Epistimio/orion/pull/886 being merged. Once it is, I'll mark this as ready for review.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -40,25 +40,25 @@ Installation
...
+Then, the repo can be installed, like so:
+
 
 .. code-block:: console
 
-    $ curl -s https://git.io/JLnCA | grep -v "^#" | xargs -n 1 -L 1 pip install
+    $ pip install git+https://github.com/Epistimio/orion.algo.robo
 
-Finally, you can install this package using PyPI
+..  TODO: Can we get this to be installable with PyPI even with links? How does Benzina do it?
```
Comment (bouthilx) [inline README.rst]: Yes, it's only the wheel that is causing issue.

File: docs/src/conf.py [hunk 2]
```diff
@@ -262,8 +277,8 @@
     (
         master_doc,
         "orion.algo.robo.tex",
-        u"Oríon algo robo Documentation",
-        u"Xavier Bouthillier",
+        "Oríon algo robo Documentation",
+        "Xavier Bouthillier",
```
Comment (bouthilx) [inline docs/src/conf.py]:
```suggestion
        "Epistímio",
```

File: setup.py [hunk 3]
```diff
@@ -1,43 +1,61 @@
...
 
-
 repo_root = os.path.dirname(os.path.abspath(__file__))
 
-tests_require = ["pytest>=3.0.0"]
+tests_require = ["pytest>=3.0.0", "pytest-mock"]
+
+extras_require = {
+    "test": tests_require,
+    # "george": [
```
Comment (bouthilx) [inline setup.py]: Where is george dep defined now?

File: src/orion/algo/robo/base.py [hunk 4]
```diff
@@ -181,129 +222,115 @@ class RoBO(BaseAlgorithm):
...
+        # a release of orion.
+        init_signature = inspect.signature(type(self).__init__)
+        self._param_names = [
+            name
+            for name, param in init_signature.parameters.items()
+            if name not in ["self", "space"]
+            and param.kind not in [param.VAR_KEYWORD, param.VAR_POSITIONAL]
+        ]
+
+    def _initialize(self, seed_rng: bool = True) -> None:
```
Comment (bouthilx) [inline src/orion/algo/robo/base.py:263]: Could this be part of `__init__` now that we don't need to wait for the transformed space to be set?

File: src/orion/algo/robo/base.py [hunk 5]
```diff
@@ -354,73 +394,57 @@ def suggest(self, num=None):
...
-                # self.optimizer.tell([point], [result])
-                return []
-
-            return [point]
+            assert self.robo is not None
+            X, y = self.XY
+            point = list(self.robo.choose_next(X, y))
+            trial = tuple_to_trial(point, self.space)
+            return [trial]
+            # TODO: Double-check, but I don't think this kind of thing is needed anymore.
```
Comment (bouthilx) [inline src/orion/algo/robo/base.py]: Yeah, I think this is not needed anymore now that the Wrapper will try algo.suggest many times if it returns many times trials that maps to the same one.

File: src/orion/algo/robo/bohamiann.py [hunk 6]
```diff
@@ -79,48 +94,50 @@ class RoBO_BOHAMIANN(RoBO):
...
-            keep_every=keep_every,
-            learning_rate=learning_rate,
-            batch_size=batch_size,
-            epsilon=epsilon,
-            mdecay=mdecay,
-            verbose=verbose,
             n_initial_points=n_initial_points,
             seed=seed,
         )
+        self.normalize_input = normalize_input
```
Comment (bouthilx) [inline src/orion/algo/robo/bohamiann.py:123]: We do they need to be set outside of `__init__` now? Won't they be missed when creating the configuration dict? (not be part of `param_names`)

File: src/orion/algo/robo/dngo.py [hunk 7]
```diff
@@ -1,14 +1,412 @@
...
+    build_kernel,
+    infer_n_hypers,
+)
 
+logger = logging.getLogger(__name__)
 
-class RoBO_DNGO(RoBO):
+
+# pylint: disable = too-many-instance-attributes
+class OrionDNGOWrapper(DNGO, WrappedRoboModel):
```
Comment (bouthilx) [inline src/orion/algo/robo/dngo.py:42]: What happened here? You needed to reimplement it?

File: setup.py (hunk 3 above)
Comment (lebrice) [reply setup.py]: It's already a requirement in RoBO

File: src/orion/algo/robo/base.py (hunk 4 above)
Comment (lebrice) [reply src/orion/algo/robo/base.py:263]: Not quite, this isn't to do with the space, it's to do with the creation of the model and optimizer and the seeding.

I'll double-check if It's possible to remove this, but if it isn't, I'll rename it to something more appropriate.

File: src/orion/algo/robo/bohamiann.py (hunk 6 above)
Comment (lebrice) [reply src/orion/algo/robo/bohamiann.py:123]: No since they are in the signature of the function, they will get picked up by the `inspect.signature(type(self).__init__)`, which works on the current class, not just the base class. 
The configuration dict keys will be all the arguments to `RoBO_BOHAMIANN.__init__`, if they are also an attribute on `self`.

File: src/orion/algo/robo/dngo.py (hunk 7 above)
Comment (lebrice) [reply src/orion/algo/robo/dngo.py:42]: Yeah, there was a "memory leak" in the train loop (accumulating the epoch losses without detaching them), and some other little things I don't quite recall atm.

File: src/orion/algo/robo/bohamiann.py (hunk 6 above)
Comment (lebrice) [reply src/orion/algo/robo/bohamiann.py:123]: Ok so here's the deal. The `seed` needs to be used to create the models, and also to set something on the models. So `seed_rng` would have to do something both before, and after the `model` and `robo` attributes are set.

File: src/orion/algo/robo/dngo.py (hunk 7 above)
Comment (lebrice) [reply src/orion/algo/robo/dngo.py:42]: Oh, yeah, I also added CUDA capability to their model.

File: src/orion/algo/robo/bohamiann.py (hunk 6 above)
Comment (lebrice) [reply src/orion/algo/robo/bohamiann.py:123]: Nevermind, forget most of what I said before, I just realized why this is here.

The problem is, `build_model` needs to be called eventually, and it may depend on some attributes that are set on `self` in subclasses. Therefore, it can't be called by the base class, but it has to be called at some point!

File: src/orion/algo/robo/base.py (hunk 4 above)
Comment (lebrice) [reply src/orion/algo/robo/base.py:263]: I just remembered why this is here.

The problem is, build_model needs to be called eventually to create the model, and the implementations in subclasses depend on some addiitonal attributes that are set on self after super().init in the subclasses. Therefore, it can't be called by the base class's init, but it has to be called at some point before we can suggest a new point. Hence the two-stage initialization.

Could this stay as-is? Or would you have any ideas on how this could be cleaner?

File: README.rst (hunk 1 above)
Comment (lebrice) [reply README.rst]: Fixed in 136103fcdb666e0461a98598c89b8daf5474aaa5

File: setup.py (hunk 3 above)
Comment (lebrice) [reply setup.py]: Removed the commented code in e4c6677fe2839a4df09ec9162268b739705cfe06

File: src/orion/algo/robo/base.py (hunk 5 above)
Comment (lebrice) [reply src/orion/algo/robo/base.py]: Fixed in f0d6e251397b40ec7ff6bfcda8579d6ae813f484

File: src/orion/algo/robo/base.py (hunk 4 above)
Comment (bouthilx) [reply src/orion/algo/robo/base.py:263]: Ok that's good. I just wanted to make sure it wasn't just because of space setter anymore.

File: src/orion/algo/robo/bohamiann.py (hunk 6 above)
Comment (bouthilx) [reply src/orion/algo/robo/bohamiann.py:123]: Oh right, I forgot about these changes. Thanks!

**Review by bouthilx — APPROVED** [review]
Thanks a lot for this update!
